chain/epoch-manager/src/shard_assignment.rs

```diff
@@ -140,21 +82,311 @@ fn assign_with_possible_repeats<T: HasStake + Eq, I: Iterator<Item = (usize, T)>
     }
 }
 
-pub trait HasStake {
-    fn get_stake(&self) -> Balance;
+/// Assigns validators to shards to satisfy `min_validators_per_shard`
+/// condition.
+/// This means that validators can be repeated.
+fn assign_to_satisfy_shards<T: HasStake + Eq + Clone>(
+    chunk_producers: Vec<T>,
+    num_shards: NumShards,
+    min_validators_per_shard: usize,
+) -> Vec<Vec<T>> {
+    let mut result: Vec<Vec<T>> = (0..num_shards).map(|_| Vec::new()).collect();
+
+    // Initially, sort by number of validators first so we fill shards up.
+    let mut shard_index: MinHeap<(usize, Balance, ShardId)> =
+        (0..num_shards).map(|s| (0, 0, s)).collect();
+
+    // Distribute chunk producers until all shards have at least the
+    // minimum requested number.  If there are not enough validators to satisfy
+    // that requirement, assign some of the validators to multiple shards.
+    let mut chunk_producers = chunk_producers.into_iter().enumerate().cycle();
+    assign_to_satisfy_shards_inner(
+        &mut shard_index,
+        &mut result,
+        &mut chunk_producers,
+        min_validators_per_shard,
+    );
+    result
 }
 
-impl HasStake for ValidatorStake {
-    fn get_stake(&self) -> Balance {
-        self.stake()
+/// Get initial chunk producer assignment for the current epoch, given the
+/// assignment for the previous epoch.
+fn get_initial_chunk_producer_assignment(
+    chunk_producers: &[ValidatorStake],
+    num_shards: NumShards,
+    prev_chunk_producers_assignment: Option<Vec<Vec<ValidatorStake>>>,
+) -> Vec<Vec<usize>> {
```

Comment:
This method basically just copies the assignment from the previous epoch right?

Reply:
Yeah, leaving only chunk producers which remained online in the current epoch.